test/python/test_job_processor.py

```diff
@@ -234,22 +218,36 @@ def test_quantum_program_online(self):
                             seed=78)
 
     def test_run_job_processor_local_parallel(self):
+        def job_done_callback(results):
+            try:
+                self.log.info(pprint.pformat(results))
+                for result, _ in results:
+                    self.assertTrue(result['result'][0]['status'] == 'DONE')
+            except Exception as e:
+                self.job_processor_exception = e
+            finally:
+                self.job_processor_finished = True
+
         njobs = 20
         job_list = []
         for i in range(njobs):
             compiled_circuit = openquantumcompiler.compile(self.qc.qasm())
             qjob = jobp.QuantumJob(compiled_circuit, backend='local_qasm_simulator')
             job_list.append(qjob)
 
-        def job_done_callback(results):
-            self.log.info(pprint.pformat(results))
-            for result, _ in results:
-                self.assertTrue(result['result'][0]['status'] == 'DONE')
-
+        self.job_processor_finished = False
+        self.job_processor_exception = None
         jp = jobp.JobProcessor(job_list, max_workers=None,
                                callback=job_done_callback)
         jp.submit(silent=True)
 
+        while not self.job_processor_finished:
```

Comment:
Do we really need to wait for the async job to finish? ... I thought moving forward to another test while the concurrent task is being executed in parallel was valid... indeed, thought that it was making testing a bit faster :)

Reply:
I think this depends on what the test is trying to test. If it's just trying to test submission, which maybe is what it was doing before, then maybe it doesn't need to wait. However, if it's trying to test that all jobs ran without error then maybe it has to wait for that. I think either we change the name of the test to test_submit... and keep the old code or accept the change. I kind of think the waiting might be ok. We'll get speed when we really run ;).

Reply:
I believe we _do_ need to wait for the callbacks to be finished - without that `while` (and the extra machinery, which admittedly is a bit clunky), `test_run_job_processor_local_parallel()` returns immediately after `jp.submit()`, and since no asserts are executed during the lifetime of that particular test (as they are _inside_ the callback) it would be always marked as `PASSED`.

So it seems we need to force those tests not to return in all cases that include assertions inside the callbacks - which I believe is the case in the four methods that have been modified (the two noted on the comments, plus two at `test_quantumprogram.py`)? For the record, the callbacks were being called, but they were called outside the lifetime of the test (they would end up showing on stdout, but not picked up correctly by the testing framework) - which was quite tricky to spot.

I'm unsure about the note about the name changes, but would be happy to rename them if needed!

Reply:
Yeah, as you said, it depends. But for the tests that are just testing the submission, the callback is set to None. In this case we are testing that local simulators can run in parallel, so we just want to make sure that all of them are executed correctly.

Reply:
For the record, @diego-plan9 showed me that waiting for all the callbacks to finish is indeed necessary. Otherwise the function will return without asserting so passing the test. When the callback is executed, the assert is evaluated and if false, an exception is thrown but the testing framework doesn't do anything with it, because the test already passed.